### DISCUSSION TITLE
Calling texture compression via C#

### DISCUSSION THREAD

**Discussion by ignarmezh** [KhronosGroup/KTX-Software#777] [Q&A] [answered]
Hi!
I'm using toktx.exe and ktx.dll for compression textures from .png to .ktx2 via cmd, with commands like:
`toktx.exe --encode etc1s --genmipmap --assign_oetf linear --assign_primaries none {newFilePath} {pathOfImage}`
Is it possible to call functions from C#? If i add dll as dependencies.
Could you help me how to do it?

**Comment by MarkCallow** [comment, marked as answer]
@ignarmezh, I assume you mean is it possible to call libktx functions from C# since toktx is an application.\*

I know very little about C# but I believe it is a language that manages all memory allocation similar to Java. If that is the case then I expect a native interface module, à la Java's JNI, is needed to enable C# programs to call functions in a native library. There is no such module in this project. If I'm not totally wrong and a module is needed, I'd be happy to accept a contribution to add one.

\* Most of what `toktx` does happens within libktx. The exceptions are scaling and resizing of images and mipmap generation.

I'm converting this to a discussion. If the discussion identifies definitively that something is needed to interface from C#, an issue can be opened to request that something.

**Reply by ignarmezh** [reply]
@MarkCallow, For that we need a `.dll` with methods which we can call through c#.
I'm not experienced enough in this matter. but as I know in c# we can add a custom method from dll with special attribute, like:
`[DllImport(@"c:\MethodNameLibrary.dll")]`
`private static extern int methodName(int b);`
`static void Main(string[] args)`
`{`
`    Console.WriteLine(methodName(3));`
`}`
About that attribute you can read [here](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.dllimportattribute?view=net-7.0)

Also, the method in the dynamic library must be declared as:
`int __declspec(dllexport) methodName(int b)`
`{`
`   return b;`
`}`

If there is an available API with a description of the methods, what they do and which parameters they accept, we could call them from C# directly I guess.

Thanks!

**Reply by MarkCallow** [reply]
I am sorry @ignarmezh. For some reason I am only just seeing your reply to my answer. When built for Windows, there is a dll, `ktx.dll` and all methods are declared with `__declspec(dllexport)`. All, except Vulkan related are declared in `ktx.h`. Vulkan ones are in `ktxvulkan.h`.

The library and headers are in the installable Windows package you can find in the project's [Releases](https://github.com/KhronosGroup/KTX-Software/releases) section. It sounds like you can add custom methods to your C# to call these functions. If you create a header for these, please submit a PR to add it to KTX-Software.

**Reply by HouleAdrien** [reply]
Hi, I made simple binding based on this discussion and my knowledge of c# /!\ it doesn't implement everything /!\ , it is just a simple c# binding that does what I wanted .
The only thing I still struggle to with is ktxTexture2_CompressBasis function that just crash/ runs infinitly in my second test of my program.
https://github.com/HouleAdrien/KtxSoftCSharpBindingExample

Edit: my issues with ktxTexture2_CompressBasis  is simply that my image seems too big (70 mo) for him and it take a lot of time to compute, but in order to use the other compress function with params I would need to upgrade my interop more, is there an other solution ?

**Reply by MarkCallow** [reply]
Please try running with a debug version of ktx.dll and see where `ktxTexture2_CompressBasis` is crashing or what loop it is stuck in.

If you need to pass params there is no alternative to calling `ktxTexture2_CompressBasisEx`.

**Reply by HouleAdrien** [reply]
Thanks, for you response , so I tried with the debug version of the dll , and it ends that it is just my file that is way too huge for fast compression but everything works, 
I will post a new version of the interop soon on my github with the ktxTexture2_CompressBasisEx function.
And also just a quick question is ktx dll thread safe ? (If I do my c# threads with call to the compression function inside them will it break ?)

Edit: Nevermind when switched back to realease dll and using ktxbasisparams it work like a charm with Huge files
